### DISCUSSION TITLE
Will this tool work with other systems? ie Orange Pi images?

### DISCUSSION THREAD

**Discussion by edel-macias-cubix** [gitbls/sdm#150] [Q&A]
I'd like to use this for other system images on more boards

**Comment by gitbls** [comment]
sdm assumes that it's operating on RasPiOS IMGs, but it doesn't check that in any way. You're certainly welcome to try it. 

If you run into problems, they may or may not be fixed depending on the nature of the problem, since I don't have an Orange Pi (or any Non-Raspberry Pi boards, for that matter) for testing.

**Reply by edel-macias-cubix** [reply]
Great, I'll test and make sure to report back with my findings. Thank you @gitbls

**Reply by gitbls** [reply]
Did you ever try this, and what were the results? Thx

**Comment by UnconnectedBedna** [comment]
I was about to ask a similar thing, that _might_ have a connection to orangepi.

The armbian image for orangepi 5 for example uses GPT as partition table, not MBR.
This creates issues if I try to use a script I maintain.

Does this tool ever change the partition table or does it simply always copy what is done on the img file?

**Comment by gitbls** [comment]
sdm only messes with partitions and partition tables under certain circumstances, and it's always done using `parted`. Here are the situations where parted is used:
* Shrinking a partition. Used to get partition size information, and also to recreate the new, smaller partition since I haven't been able to get `parted resizepart` to work as desired
* Extending the rootfs partition
* During the process of enabling rootfs encryption to get the size of the root partition
* By the `parted` plugin when adding a new partition

In all of the above operations sdm does not care at all about the partition table type, it only looks at the information on the partitions themselves.

I think the bigger question is whether the operations that sdm does to fulfill your requests will work on your non-RasPiOS. sdm assumes that any files it uses are in the locations as they are on Debian, which is what RasPiOS is based on. It also assumes `raspi-config` is available for certain operations, but these should be pretty clear.

IOW if your distro is different from RasPiOS but you don't ask sdm to do anything that would uncover those differences, it should *just work*, as long as the host OS on which you're running sdm has these packages: systemd-container qemu-user-static binfmt-support parted

It's easy enough to try sdm. Installing it doesn't spew stuff around the system, it's all in /usr/local/sdm, plus the link /usr/local/bin/sdm which points to /usr/local/sdm. 

Why don't you simply give it a try?

EDIT: I just downloaded Armbian_23.11.1_Orangepi5_bookworm_legacy_5.10.160.img.xz, and it indeed has a GPT partition table:
```
pw/ssdy/work# sfdisk -l Armbian_23.11.1_Orangepi5_bookworm_legacy_5.10.160.img
Disk Armbian_23.11.1_Orangepi5_bookworm_legacy_5.10.160.img: 2.45 GiB, 2634022912 bytes, 5144576 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: 59752424-0B97-8C4D-9A2A-C1A6B85E8161

Device                                                   Start     End Sectors  Size Type
Armbian_23.11.1_Orangepi5_bookworm_legacy_5.10.160.img1  32768  557055  524288  256M Linux extended boot
Armbian_23.11.1_Orangepi5_bookworm_legacy_5.10.160.img2 557056 5144542 4587487  2.2G Linux filesystem
pw/ssdy/work# sdm --explore Armbian_23.11.1_Orangepi5_bookworm_legacy_5.10.160.img
* Enter IMG 'Armbian_23.11.1_Orangepi5_bookworm_legacy_5.10.160.img'
root@sdm:/# ls
bin  boot  dev  etc  home  lib  lib64  lost+found  media  mnt  opt  proc  root  run  sbin  selinux  srv  sys  tmp  usr  var
root@sdm:/# 
root@sdm:/# apt update
Get:2 http://security.debian.org bookworm-security InRelease [48.0 kB]
Get:3 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:1 http://armbian.tnahosting.net/apt bookworm InRelease [53.3 kB]                       
Get:4 http://security.debian.org bookworm-security/main arm64 Packages [132 kB]
Get:5 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]              
Get:6 http://deb.debian.org/debian bookworm-backports InRelease [56.5 kB]
Get:7 http://armbian.tnahosting.net/apt bookworm/bookworm-utils all Packages [10.7 kB]
Get:8 http://armbian.tnahosting.net/apt bookworm/bookworm-utils arm64 Packages [24.5 kB]
Get:9 http://armbian.tnahosting.net/apt bookworm/bookworm-utils all Contents (deb) [9641 B]
Get:10 http://armbian.tnahosting.net/apt bookworm/bookworm-utils arm64 Contents (deb) [16.7 kB]
Get:11 http://deb.debian.org/debian bookworm/main arm64 Packages [11.9 MB]                 
Get:12 http://armbian.tnahosting.net/apt bookworm/bookworm-desktop arm64 Packages [3045 B]
Get:13 http://armbian.tnahosting.net/apt bookworm/bookworm-desktop arm64 Contents (deb) [162 kB]
Get:14 http://deb.debian.org/debian bookworm-updates/main arm64 Packages.diff/Index [9483 B]            
Get:15 http://deb.debian.org/debian bookworm-updates/main arm64 Contents (deb).diff/Index [7239 B]
Get:16 http://deb.debian.org/debian bookworm-updates/main all Contents (deb).diff/Index [8361 B]
Get:17 http://deb.debian.org/debian bookworm-backports/main arm64 Packages.diff/Index [63.3 kB]
Ign:17 http://deb.debian.org/debian bookworm-backports/main arm64 Packages.diff/Index
Get:18 http://deb.debian.org/debian bookworm-backports/main all Contents (deb).diff/Index [63.6 kB]
Get:19 http://deb.debian.org/debian bookworm-backports/main arm64 Contents (deb).diff/Index [63.4 kB]
Get:20 http://deb.debian.org/debian bookworm-backports/contrib arm64 Packages.diff/Index [17.3 kB]
Get:21 http://deb.debian.org/debian bookworm-backports/contrib arm64 Contents (deb).diff/Index [6117 B]
Get:22 http://deb.debian.org/debian bookworm-backports/contrib all Contents (deb).diff/Index [10.6 kB]
Get:23 http://deb.debian.org/debian bookworm-backports/non-free arm64 Packages.diff/Index [2732 B]
Get:24 http://deb.debian.org/debian bookworm-backports/non-free all Contents (deb).diff/Index [2732 B]
Get:25 http://deb.debian.org/debian bookworm-updates/main arm64 Packages T-2023-12-29-1403.39-F-2023-12-15-1408.04.pdiff [7212 B]
Get:25 http://deb.debian.org/debian bookworm-updates/main arm64 Packages T-2023-12-29-1403.39-F-2023-12-15-1408.04.pdiff [7212 B]
Get:26 http://deb.debian.org/debian bookworm-updates/main arm64 Contents (deb) T-2023-12-29-1403.39-F-2023-12-15-1408.04.pdiff [229 kB]
Get:26 http://deb.debian.org/debian bookworm-updates/main arm64 Contents (deb) T-2023-12-29-1403.39-F-2023-12-15-1408.04.pdiff [229 kB]
Get:27 http://deb.debian.org/debian bookworm-updates/main all Contents (deb) T-2023-12-29-1403.39-F-2023-12-15-1408.04.pdiff [207 kB]
Get:27 http://deb.debian.org/debian bookworm-updates/main all Contents (deb) T-2023-12-29-1403.39-F-2023-12-15-1408.04.pdiff [207 kB]
Get:28 http://deb.debian.org/debian bookworm-backports/main all Contents (deb) T-2024-01-28-0206.35-F-2023-11-30-0808.24.pdiff [506 kB]
Get:28 http://deb.debian.org/debian bookworm-backports/main all Contents (deb) T-2024-01-28-0206.35-F-2023-11-30-0808.24.pdiff [506 kB]
Get:29 http://deb.debian.org/debian bookworm-backports/main arm64 Contents (deb) T-2024-01-28-0804.50-F-2023-11-30-2009.37.pdiff [193 kB]
Get:29 http://deb.debian.org/debian bookworm-backports/main arm64 Contents (deb) T-2024-01-28-0804.50-F-2023-11-30-2009.37.pdiff [193 kB]
Get:30 http://deb.debian.org/debian bookworm-backports/contrib arm64 Packages T-2024-01-25-0204.30-F-2023-12-01-0205.30.pdiff [2332 B]
Get:30 http://deb.debian.org/debian bookworm-backports/contrib arm64 Packages T-2024-01-25-0204.30-F-2023-12-01-0205.30.pdiff [2332 B]
Get:31 http://deb.debian.org/debian bookworm-backports/contrib arm64 Contents (deb) T-2024-01-01-0204.24-F-2023-12-04-1405.26.pdiff [2085 B]
Get:31 http://deb.debian.org/debian bookworm-backports/contrib arm64 Contents (deb) T-2024-01-01-0204.24-F-2023-12-04-1405.26.pdiff [2085 B]
Get:32 http://deb.debian.org/debian bookworm-backports/contrib all Contents (deb) T-2024-01-25-0204.30-F-2023-12-04-1405.26.pdiff [6303 B]
Get:32 http://deb.debian.org/debian bookworm-backports/contrib all Contents (deb) T-2024-01-25-0204.30-F-2023-12-04-1405.26.pdiff [6303 B]
Get:33 http://deb.debian.org/debian bookworm-backports/non-free arm64 Packages T-2023-12-20-2024.36-F-2023-12-20-2024.36.pdiff [256 B]
Get:33 http://deb.debian.org/debian bookworm-backports/non-free arm64 Packages T-2023-12-20-2024.36-F-2023-12-20-2024.36.pdiff [256 B]
Get:34 http://deb.debian.org/debian bookworm-backports/non-free all Contents (deb) T-2023-12-20-2024.36-F-2023-12-20-2024.36.pdiff [19.7 kB]
Get:34 http://deb.debian.org/debian bookworm-backports/non-free all Contents (deb) T-2023-12-20-2024.36-F-2023-12-20-2024.36.pdiff [19.7 kB]
Get:35 http://deb.debian.org/debian bookworm-backports/contrib arm64 Packages T-2024-01-25-0204.30-F-2024-01-25-0204.30.pdiff [723 B]
Get:35 http://deb.debian.org/debian bookworm-backports/contrib arm64 Packages T-2024-01-25-0204.30-F-2024-01-25-0204.30.pdiff [723 B]
Get:36 http://deb.debian.org/debian bookworm/main all Contents (deb) [33.0 MB]
Get:37 http://deb.debian.org/debian bookworm/main arm64 Contents (deb) [11.2 MB]                                                                                         
Get:38 http://deb.debian.org/debian bookworm/contrib arm64 Packages [54.7 kB]                                                                                            
Get:39 http://deb.debian.org/debian bookworm/contrib arm64 Contents (deb) [37.7 kB]                                                                                      
Get:40 http://deb.debian.org/debian bookworm/non-free arm64 Packages [94.4 kB]                                                                                           
Get:41 http://deb.debian.org/debian bookworm/non-free arm64 Contents (deb) [43.4 kB]                                                                                     
Get:42 http://deb.debian.org/debian bookworm/non-free-firmware arm64 Packages [6401 B]                                                                                   
Get:43 http://deb.debian.org/debian bookworm/non-free-firmware arm64 Contents (deb) [243 B]                                                                              
Get:44 http://deb.debian.org/debian bookworm-backports/main arm64 Packages [168 kB]                                                                                      
Get:45 http://deb.debian.org/debian bookworm-updates/main arm64 Contents (deb) [253 kB]                                                                                  
Fetched 58.9 MB in 10s (5999 kB/s)                                                                                                                                       
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
30 packages can be upgraded. Run 'apt list --upgradable' to see them.
N: Repository 'http://deb.debian.org/debian bookworm InRelease' changed its 'Version' value from '12.2' to '12.4'
```
As demonstrated above, sdm can operate on the IMG. Please give it a workout and LMK how it goes.

**Reply by UnconnectedBedna** [reply]
> since I haven't been able to get parted resizepart to work as desired

I might have some insight to that.
I do not know why, but using resizepart does not work "the normal way", the `--script` option seems to be ignored for some reason, I have tried EVERYTHING!
The only thing that actually work is doing it the hacky way using `---pretend-input-tty`
Example:
```
printf 'Yes\n' | parted -a none "$LOOP" unit B resizepart "$IMG_ROOT_PARTN" "$TRUNCATE_TOTAL" ---pretend-input-tty 2>&1
```

LOOP = device path (I have the img file looped, hence my path is something like `/dev/loop0`
IMG_ROOT_PARTN = kinda self explanatory, the partition number, usually 2 if you have a boot partition
TRUNCATE_TOTAL = total size in bytes
---pretend-input-tty = The only hint I can find about this option is that it is probably a leftover from the devs of parted, not mentioned in the manual. But I assume it takes the printf, types "yes" and the newline is just an enter press.

When I expand partitions I just remove the partition, then recreate it and use `expand2fs` to expand the filesystem.

> Why don't you simply give it a try?

You know what, how about "if you try mine, I try yours". xD (I am one of those guys that just throws in a ssh file and a userconf.txt on the img file and I'm done)
Allthough, if you are curious about the btrfs stuff I just implemented, try the testing branch.

**Edit**
But to expand on my initial thing, things also change if partition table is gpt instead of mbr for me, because how I create the img file, when gpt is in use `parted` no longer accepts me just force deleting the root partition, neither does `sfdisk`, I had to rely on `gdisk`.

**Edit 2**
> As demonstrated above, sdm can operate on the IMG. Please give it a workout and LMK how it goes.

I would love to help out, but I do not own a opi5 either. :cry: 
@jsm174

**Reply by gitbls** [reply]
I believe I'm having a different issue with resizepart than you might have seen. The error I'm getting is
```
pw/ssdy/work# sdm --shrink 2023-12-05-raspios-bookworm-arm64-lite.img 
* Start shrink on IMG '2023-12-05-raspios-bookworm-arm64-lite.img'
> Check the file system
rootfs: 89869/703456 files (0.2% non-contiguous), 1127148/2895872 blocks
resize2fs 1.47.0 (5-Feb-2023)
> Shrink the file system from 2895872 to 1238956 4096-byte blocks
resize2fs 1.47.0 (5-Feb-2023)
Resizing the filesystem on /dev/loop0 to 1238956 (4k) blocks.
Begin pass 3 (max = 89)
Scanning inode table          XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
The filesystem on /dev/loop0 is now 1238956 (4k) blocks long.

BYT;
/dev/loop0:11861491712B:loopback:512:512:loop:Loopback device:;
1:0B:11861491711B:11861491712B:ext4::;
> Resize partition: 5615828991
Error: The location 5615828991 is outside of the device /dev/loop0.
? parted resize partition failed 1
```
I will investigate further, and if your solution is THE solution, I'll use it, but I'm not overly optimistic.

So, you asked about the Orange pi5 but don't have one? Please enlighten me as to the purpose of your query.

**Reply by UnconnectedBedna** [reply]
This was a while ago but:
> Error: The location 5615828991 is outside of the device /dev/loop0.

I think that was the exact error I also got and could not figure out why. Could also be I relate it to another error in my head, so take my info with a grain of sand.
But before you go hacking around with ---pretend-hack-parted, I do not know what the exact parted line you use, but have you tired to set alignement to none? ie: `parted -a none yadayadayada`

Also, the bytes are not the same in your BYT data.
I remember going crazy with using resize2fs an blocks, since the boot is 512, and the rest is 4K, so I changed it to kilobytes (or kibibytes or whatever they are called, I'm old AF) and used `resize2fs -p -f "$IMG_DEV_ROOT_PATH" "$TOTALK"K 2>&1`. The trailing K tells it to use kilobytes. (with 1024, I guess KiB)

ALSO!!
Remember the size used in parted is NOT the same as with resize2fs, you need to include the boot as well, parted is asking for the end byte, not the size byte.
And again, remember, the boot and the space before the boot is calculated using 512, everythng after is 4096!
I suspect you calculated everything with 4096 and then it becomes too big. :)

**Edit**
Another memory popped into my brain.
I used to just setup the loops using an offset, and then try to use parted to resize the partion. In my head, it should not matter anyway right?
WRONG!
Mount the device (img) directly to the loop and then partprobe it, you will then see `/dev/loop0p1` and `/dev/loop0p2` instead of just `/dev/loop`

You still run parted directly towards `/dev/loop0` and reference the partition number instead (like I did in the ---pretend-tty hack).

Just an extra thought since I do not know how you loop it.

> So, you asked about the Orange pi5 but don't have one? Please enlighten me as to the purpose of your query.

2 Reasons.
1. An issue was created on my github.
2. Your application is big enough for me to consider changing things IF it were to create GPT in other situations than above.

Both of above are now solved (for me at least, but maybe YOU appreciate the heads up). :)

**Reply by gitbls** [reply]
OK, got shrink working using resizepart. I had a couple of errors, but once I fixed those found that I needed your ` ---pretend-input-tty` trick. *Thanks for that!*

Update coming this week with this and several other improvements.

**Reply by UnconnectedBedna** [reply]
You are more than welcome.

Now you also know of a script that does not care what os is in use.
Feel free to use it if you venture outside of the rpi software space, like armbian f.ex. :D
(I tried the script on a running installation of manjaro kde, it worked even on a system like that so not even just sbc:s) :smile:
